src/databricks/labs/remorph/reconcile/compare.py

```diff
@@ -0,0 +1,86 @@
+from pyspark.sql import DataFrame
+from pyspark.sql.functions import col, expr, lit
+
+from databricks.labs.remorph.reconcile.constants import Constants
+from databricks.labs.remorph.reconcile.recon_config import ReconcileOutput
+
+
+def reconcile_data(source: DataFrame, target: DataFrame, join_columns: list[str], report_type: str) -> ReconcileOutput:
+    source_alias = 'src'
+    target_alias = 'tgt'
+    if report_type not in {"data", "all"}:
+        join_columns = Constants.hash_column_name
+    df = source.alias(source_alias).join(other=target.alias(target_alias), on=join_columns, how="full")
+
+    mismatch = _get_mismatch_data(df, source_alias, target_alias) if report_type in {"all", "data"} else None
+    missing_in_src = (
+        df.filter(col(f"{source_alias}.{Constants.hash_column_name}").isNull())
+        .select((join_columns if report_type == "all" else f"{target_alias}.*"))
+        .drop(Constants.hash_column_name)
+    )
+    missing_in_tgt = (
+        df.filter(col(f"{target_alias}.{Constants.hash_column_name}").isNull())
+        .select((join_columns if report_type == "all" else f"{source_alias}.*"))
+        .drop(Constants.hash_column_name)
+    )
+    return ReconcileOutput(missing_in_src=missing_in_src, missing_in_tgt=missing_in_tgt, mismatch=mismatch)
+
+
+def _get_mismatch_data(df: DataFrame, src_alias: str, tgt_alias: str) -> DataFrame:
+    return (
+        df.filter(
+            (col(f"{src_alias}.{Constants.hash_column_name}").isNotNull())
+            & (col(f"{tgt_alias}.{Constants.hash_column_name}").isNotNull())
+        )
+        .withColumn(
+            "hash_match",
+            col(f"{src_alias}.{Constants.hash_column_name}") == col(f"{tgt_alias}.{Constants.hash_column_name}"),
+        )
+        .filter(col("hash_match") == lit(False))
+        .select(f"{src_alias}.*")
+        .drop(Constants.hash_column_name)
+    )
+
+
+def capture_mismatch_data_and_columns(
+    source: DataFrame, target: DataFrame, key_columns: list[str]
+) -> (DataFrame, list[str]):
+    source_columns = source.columns
+    target_columns = target.columns
+
+    if source_columns != target_columns:
+        message = "source and target should have same columns for capturing the mismatch data"
+        raise ValueError(message)
```

Comment:
can you raise a custom Error saying column name mismatch and which column name is mismatched